### PR TITLE
ath79: add support for D-link DAP-1720 A1

### PR DESCRIPTION
**Pull request by raenye**
D-Link DAP-1720 rev A1 is a mains-powered AC1750 Wi-Fi range extender, manufactured by Alpha Networks [**8WAPAC28.1A1G**].
(in square brackets: PCB silkscreen markings)

Specifications:
* CPU (Qualcomm Atheros QCA9563-AL3A [**U5**]):
  775 MHz single core MIPS 74Kc;
* RAM (Winbond W9751G6KB-25J [**U3**]):
  64 MiB DDR2;
* ROM (Winbond W25Q128FV [**U16**]):
  16 MiB SPI NOR flash;
* Ethernet (AR8033-AL1A PHY [**U1**], no switch):
  1 GbE RJ45 port (no PHY LEDs);
* Wi-Fi
  * 2.4 GHz (Qualcomm Atheros QCA9563-AL3A [**U5**]):
    3x3 802.11n;
  * 5 GHz (Qualcomm Atheros QCA9880-BR4A [**U9**]):
    3x3 802.11ac Wave 1;
  * 3 foldable dual-band antennas (U.fl) [**P1**],[**P2**],[**P3**];
* GPIO LEDs:
  * RSSI low (red/green) [**D2**];
  * RSSI medium (green) [**D3**];
  * RSSI high (green) [**D4**];
  * status (red/green) [**D5**];
* GPIO buttons:
  * WPS [**SW1**], co-located with status LED;
  * reset [**SW4**], accessible via hole in the side;
* Serial/UART: Tx-Gnd-3v3-Rx [**JP1**], Tx is the square pin, 1.25mm pitch;
  125000-8-n-1 in U-boot, 115200-8-n-1 in kernel;
* Misc:
  * 12V VCC [**JP2**], fed from internal 12V/1A AC to DC converter;
  * on/off slide switch [**SW2**] (disconnects VCC mechanically);
  * unpopulated footprints for a Wi-Fi LED [**D1**];
  * unpopulated footprints for a 4-pin 3-position slide switch [**SW3**];

MAC addresses:
* Label = LAN;
* 2.4 GHz WiFi = LAN;
* 5 GHz WiFi = LAN+2;

Installation:
* `factory.bin` can be used to install OpenWrt from OEM firmware via the standard upgrade webpage at http://192.168.0.50/UpdateFirmware.html
* `recovery.bin` can be used to install OpenWrt (or revert to OEM firmware) from D-Link Web Recovery. To enter web recovery, keep reset button pressed and then power on the device. Reset button can be released when the red status LED is bright; it will then blink slowly. Set static IP to 192.168.0.10, navigate to http://192.168.0.50 and upload 'recovery.bin'. Note that in web recovery mode the device ignores ping and DHCP requests.

Note: 802.11s is not supported by the default `ath10k` driver and firmware, but is supported by the non-CT driver and firmware variants. The `-smallbuffers` driver variant is recommended due to RAM size.

Co-developed-by: @protectivedad

### CODE REVIEW

**Comment by raenye** [comment]
If #13930 is merged before this, I'll update the makefile to create also `factory.bin`.

File: target/linux/ath79/dts/qca9563_dlink_dap-1720-a1.dts [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+	};
+};
+
+&wmac {
+	status = "okay";
+
+	nvmem-cells = <&cal_ath9k>, <&macaddr_ath9k 0>;
+	nvmem-cell-names = "calibration", "mac-address";
+};
+
```
Comment (neheb) [inline target/linux/ath79/dts/qca9563_dlink_dap-1720-a1.dts:215]: extra newline?

Comment (raenye) [reply target/linux/ath79/dts/qca9563_dlink_dap-1720-a1.dts:215]: Did you mean at the EOF? I'll remove it.

**Comment by raenye** [comment]
@neheb maybe you know why my DTS makes `dmesg` spit a handful of the following warning:
```
OF: Bad cell count for /ahb/spi@1f000000/flash@0/partitions
```

**Comment by neheb** [comment]
No idea. Might be a problem with ath79 dtsi files. Not sure.

File: target/linux/ath79/dts/qca9563_dlink_dap-1720-a1.dts [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+		};
+	};
+};
+
+&mdio0 {
+	status = "okay";
+
+	phy0: ethernet-phy@0 {
+		reg = <0>;
+		reset-gpios = <&gpio 11 GPIO_ACTIVE_LOW>; // necessary?
```
Comment (hauke) [inline target/linux/ath79/dts/qca9563_dlink_dap-1720-a1.dts:183]: If this is the reset GPIO for the PHY please keep it in. Sometimes it is not always necessary  to reset the PHY and it is already in the right state. The generic Linux PHY framework should take care.

File: target/linux/ath79/dts/qca9563_dlink_dap-1720-a1.dts [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+			};
+
+			partition@70000 {
+				compatible = "seama";
+				label = "firmware";
+				reg = <0x070000 0xf80000>;
+			};
+
+			partition@ff0000 {
+				label = "art"; // vendor calls it `radiocfg`
```
Comment (hauke) [inline target/linux/ath79/dts/qca9563_dlink_dap-1720-a1.dts:156]: Why don't you use the names the vendor used for the partitions OpenWrt does not directly access?

File: target/linux/ath79/dts/qca9563_dlink_dap-1720-a1.dts [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+		compatible = "gpio-leds";
+
+		led_status_red: led-1 {
+			color = <LED_COLOR_ID_RED>;
+			function = LED_FUNCTION_STATUS;
+			gpios = <&gpio 5 GPIO_ACTIVE_LOW>;
+			panic-indicator;
+		};
+
+		led_status_green: led-2 {
```
Comment (hauke) [inline target/linux/ath79/dts/qca9563_dlink_dap-1720-a1.dts:50]: Is it common in the Linux kernel to just give the LEDs numbers in the dts?

I haven't followed the LED definition in DTS in Linux lately.

**Comment by hauke** [comment]
What does the `[U5]` and so on in your commit description mean?

**Comment by raenye** [comment]
> What does the [U5] and so on in your commit description mean?

PCB silkscreen text.

File: target/linux/ath79/dts/qca9563_dlink_dap-1720-a1.dts (hunk 4 above)
Comment (raenye) [reply target/linux/ath79/dts/qca9563_dlink_dap-1720-a1.dts:50]: I've seen both sequential (`led-1`, `led-2`, ...) and something like `led-wan`, `led-status`. Not sure what is preferred in OpenWrt.

File: target/linux/ath79/dts/qca9563_dlink_dap-1720-a1.dts (hunk 3 above)
Comment (raenye) [reply target/linux/ath79/dts/qca9563_dlink_dap-1720-a1.dts:156]: `art`, `uboot`, etc seemed more common. It seems to only matter within the DTS.
Should I change?

File: target/linux/ath79/dts/qca9563_dlink_dap-1720-a1.dts (hunk 2 above)
Comment (raenye) [reply target/linux/ath79/dts/qca9563_dlink_dap-1720-a1.dts:183]: Will fix. Thanks for your review.

File: target/linux/ath79/dts/qca9563_dlink_dap-1720-a1.dts (hunk 4 above)
Comment (hauke) [reply target/linux/ath79/dts/qca9563_dlink_dap-1720-a1.dts:50]: I prefer names like `led-wan`. This makes it easier to read the dts file for humans.

Comment (raenye) [reply target/linux/ath79/dts/qca9563_dlink_dap-1720-a1.dts]: Done.

**Comment by neheb** [comment]
@rayne if that OF warning still appears,

```
#address-cells = <1>;
#size-cells = <1>;
```

is needed before partitions under flash.

**Comment by raenye** [comment]
> is needed before partitions under flash.

Thanks @neheb. Is this needed both within `flash@0` and within `partitions`?

P.S. you pinged Rayne; I'm raenye :)

**Comment by neheb** [comment]
Just the former. Apparently that warning gets triggered when the parent is missing those.

**Comment by raenye** [comment]
@neheb I see I have this warning on my `mt7621` device too.
In fact, it seems *all `mt7621` devices* have it wrong.
Should I (in a separate PR) fix all DTs using a NOR* flash?

*For some reason I don't get this warning on Xiaomi Redmi AX6S `mt7622` which has a NAND flash.

**Comment by neheb** [comment]
Feel free to do so. Although @mpratt14 changes it to a debug warning instead of an error. Given that it’s not upstream, makes sense to fix it.

**Comment by mpratt14** [comment]
yeah, "bad cell count" from lack of defining address-cells and size-cells in the partitions is in the entirety of openwrt, in all subtargets

the reason is, partitions and their sub nodes are only just recently having address translation because of the addition of of_platform_populate() being used in mtdcore, and that was done to make the nvmem driver capable of using the partitions nodes as nvmem providers (which are usually standalone devices)

basically, the kernel is parsing the data in the nodes *as if* they were real physical devices in the SOC with a bus address range and registers and so on, but they aren't, the nodes are literally just info for the virtual flash space. the error is meant to tell the developer why the device isn't working in the case of the dts being written wrong, but it doesn't mean anything for nodes not representing a real device.

**Comment by raenye** [comment]
> it doesn't mean anything for nodes not representing a real device.

I prefer less warnings :)

**Comment by mpratt14** [comment]
> I prefer less warnings :)

same here, that's why I was hoping my PR with a commit that silences it gets accepted before it's 1 year old zzzzzz

maybe I should put that in a separate PR by itself...

**Comment by raenye** [comment]
> my PR with a commit that silences it gets accepted

Which one?

File: target/linux/ath79/dts/qca9563_dlink_dap-1720-a1.dts (hunk 3 above)
Comment (hauke) [reply target/linux/ath79/dts/qca9563_dlink_dap-1720-a1.dts:156]: ok fine with me.

**Comment by neheb** [comment]
@raenye here's the kicker.

it fixes a runtime warning but introduces a compile time warning(currently suppressed) saying that it's unnecessary. Confusingly, ath79 throws the runtime warning but mt7621 does not. It's some very strange dts interaction.

**Comment by mpratt14** [comment]
what's the compile time warning say?

**Comment by neheb** [comment]
It’s the bad cell count one.
